### ISSUE TITLE
Bug in fine-tuning for Qwen2.5-VL with LoRA

### ISSUE THREAD

**Issue by Bytes-Lin** [huggingface/peft#2880]
I have set the lora config as follows to fine-tuning Qwen2.5-VL 3B model.

<img width="599" height="493" alt="Image" src="https://github.com/user-attachments/assets/1a84f1d2-a633-4bd7-b9cc-de737ef726e5" />

I wanted to update the weights of qkv module in ViT of model, but I found that although the attribute(requires_grad) of parameters is True, the gradient is 0 while training, meaning that the parameters of qkv module in ViT have not been trained correctly.

Later, I manually set the input of ViT requires gradient, then I found the training works. Also, I tried full-parameter training, the weights of ViT updated well.

So, I wonder that is this a bug in PEFT or just happens in the training of MLLMs.

**Comment by githubnemo** [comment]
Can you perhaps supply a small reproducer? To clarify, the `qkv` module wouldn't receive gradients as it is not trained, the adapter is trained. Can you confirm that the adapter doesn't receive gradients?

**Comment by Bytes-Lin** [comment]
> Can you perhaps supply a small reproducer? To clarify, the `qkv` module wouldn't receive gradients as it is not trained, the adapter is trained. Can you confirm that the adapter doesn't receive gradients?

I have a test demo for you to reproduce the bug, for convenience, the data is mocked at runing time. You need to change the path for load and save the model weights.
By the way, the version of transformers, peft and pytorch are as follows.

<img width="516" height="1267" alt="Image" src="https://github.com/user-attachments/assets/f1eae608-7f74-4ec1-b350-53f1f9096467" />

Here is the demo.

[sample.py](https://github.com/user-attachments/files/23249052/sample.py)


While training, the grad is 0 all the time, and the weights of qkv module before training is as the same as the weights of qkv module atfer training.

<img width="1091" height="148" alt="Image" src="https://github.com/user-attachments/assets/f709d68c-c3bf-44cb-9ce5-60d64b370f97" />

**Comment by BenjaminBossan** [comment]
Thanks for providing the reproducer. When I tested it locally, the issue stemmed from gradient checkpointing. When setting `gradient_checkpointing=False`, there were gradients. Could you please give this a try?

I also checked if it is PEFT related and it was not: When I fully fine-tune the `qkv` layer directly, without PEFT, I have the same issue with gradient checkpointing. It's probably a problem in transformers. Note that calling `model.enable_input_require_grads()` makes no difference.

**Comment by Bytes-Lin** [comment]
> Thanks for providing the reproducer. When I tested it locally, the issue stemmed from gradient checkpointing. When setting `gradient_checkpointing=False`, there were gradients. Could you please give this a try?
> 
> I also checked if it is PEFT related and it was not: When I fully fine-tune the `qkv` layer directly, without PEFT, I have the same issue with gradient checkpointing. It's probably a problem in transformers. Note that calling `model.enable_input_require_grads()` makes no difference.

Thanks for replying. I changed the `gradient_checkpointing=False` in Lora fine-tuning, and it works.

<img width="1246" height="154" alt="Image" src="https://github.com/user-attachments/assets/2c7b6fb5-d693-4f61-8ffa-905e73869577" />

But when I tried with full fine-tuning, I set `gradient_checkpointing=True`, the gradient updated normally and qkv weights can be trained.

Actually, at first, I trained all the gate_proj, up_proj and down_proj modules in Qwen2.5-VL. Because the same MLP structure is employed in ViT and LLM of Qwen2.5-VL model, so in theory, the weights of all the gate_proj, up_proj and down_proj modules should be updated, but accidentally I find the weights of modules in ViT didn't update.
Munually I set `hidden_states.requires_grad = True` in ViT, in this case, trainable parameters in ViT could update properly.

<img width="1227" height="289" alt="Image" src="https://github.com/user-attachments/assets/7be24590-0978-470b-b60c-52c85de5deac" />

So I can't ensure whether this issue is caused by transformers or peft, but it's really awful.

**Comment by BenjaminBossan** [comment]
Thanks for testing and confirming. Here is my slightly modified script:

```python
import random

import torch
from torchvision.transforms import transforms
from datasets import Dataset

from qwen_vl_utils import process_vision_info
from transformers import (
    TrainingArguments,
    Trainer,
    DataCollatorForSeq2Seq,
    Qwen2_5_VLForConditionalGeneration,
    Qwen2_5_VLProcessor,
    AutoTokenizer,
)


USE_PEFT = False
GRADIENT_CHECKPOINTING = True


def process_func_batch(examples):
    MAX_LENGTH = 1024
    input_ids, attention_mask, labels, pixel_values, image_grid_thw = [], [], [], [], []
    for example in zip(examples["question"], examples["assistant_answer"]):
        input_content, output_content = example
        messages = [
            {
                "role": "user",
                "content": [
                    {"type": "image", "image": transforms.ToPILImage()(torch.randn(280, 280))},
                    {"type": "text", "text": input_content},
                ],
            }
        ]
        text = processor.apply_chat_template(
            messages, tokenize=False, add_generation_prompt=True
        )
        image_inputs, video_inputs = process_vision_info(messages)
        inputs = processor(
            text=[text],
            images=image_inputs,
            videos=video_inputs,
            padding=False,
            return_tensors="pt",
        )

        inputs_dict = {key: value.tolist() for key, value in inputs.items()}
        instruction_input_ids = inputs_dict['input_ids'][0]
        instruction_attention_mask = inputs_dict['attention_mask'][0]

        response = tokenizer(output_content, add_special_tokens=False)
        response_input_ids = response['input_ids']
        response_attention_mask = response['attention_mask']

        remaining_length = MAX_LENGTH - len(instruction_input_ids) - 1
        if remaining_length < 0:
            truncation_length = len(instruction_input_ids) + remaining_length
            instruction_input_ids = instruction_input_ids[:truncation_length]
            instruction_attention_mask = instruction_attention_mask[:truncation_length]
            remaining_length = 0

        current_input_ids = (
                instruction_input_ids + response_input_ids[:remaining_length] + [tokenizer.pad_token_id]
        )
        current_attention_mask = (
                instruction_attention_mask + response_attention_mask[:remaining_length] + [1]
        )
        current_labels = (
                [-100] * len(instruction_input_ids) +
                response_input_ids[:remaining_length] +
                [tokenizer.pad_token_id]
        )

        if len(current_input_ids) < MAX_LENGTH:
            current_input_ids += [tokenizer.pad_token_id] * (MAX_LENGTH - len(current_input_ids))
            current_attention_mask += [0] * (MAX_LENGTH - len(current_attention_mask))
            current_labels += [-100] * (MAX_LENGTH - len(current_labels))

        input_ids.append(current_input_ids)
        attention_mask.append(current_attention_mask)
        labels.append(current_labels)
        pixel_values.append(inputs_dict['pixel_values'])
        image_grid_thw.append(torch.tensor(inputs_dict['image_grid_thw']).squeeze(0))

    return {
        "input_ids": torch.tensor(input_ids),
        "attention_mask": torch.tensor(attention_mask),
        "labels": torch.tensor(labels),
        "pixel_values": torch.tensor(pixel_values, requires_grad=True),
        "image_grid_thw": torch.stack(image_grid_thw)
    }


if __name__ == '__main__':
    random.seed(42)
    torch.manual_seed(42)
    torch.cuda.manual_seed(42)

    dtype = torch.bfloat16
    model_id = 'Qwen/Qwen2.5-VL-3B-Instruct'
    model_save_path = '/tmp/model/test/'

    tokenizer = AutoTokenizer.from_pretrained(pretrained_model_name_or_path=model_id)
    processor = Qwen2_5_VLProcessor.from_pretrained(pretrained_model_name_or_path=model_id)
    model = Qwen2_5_VLForConditionalGeneration.from_pretrained(model_id, device_map='auto', dtype=dtype)
    model.enable_input_require_grads()
    model.config.use_cache = False

    data_list = [
        {'question': 'Which side shows more severe pleural effusion? (right, left, same severity, none)',
         'assistant_answer': 'absent'},
        {'question': 'Which side shows more severe pulmonary opacities? (right, left, same severity, none)',
         'assistant_answer': 'right'},
        {'question': 'Which side shows more severe atelectasis? (right, left, same severity, none)',
         'assistant_answer': 'same severity'},
        {'question': 'Are the lung apex and costodiaphragmatic angle fully visualized on both sides? (yes or no)',
         'assistant_answer': 'yes'},
        {'question': 'Is the X-ray image malrotated? (yes or no)', 'assistant_answer': 'no'},
        {'question': 'Does the chest X-ray show cardiomegaly? (yes or no)', 'assistant_answer': 'yes'},
    ]

    dataset = Dataset.from_list(data_list)
    train_dataset = dataset.map(process_func_batch, batched=True, batch_size=1, num_proc=1,
                                remove_columns=['question', 'assistant_answer'])

    model.train()
    model.enable_input_require_grads()

    if USE_PEFT:
        from peft import LoraConfig, get_peft_model, TaskType
        config = LoraConfig(
            task_type=TaskType.CAUSAL_LM,
            target_modules=["qkv"],
            init_lora_weights=False,
            inference_mode=False,
            r=16,
            lora_alpha=32,
            lora_dropout=0.1,
        )
        peft_model = get_peft_model(model, config)
        peft_model.print_trainable_parameters()
    else:
        # fully fine-tune qkv layers
        peft_model = model
        peft_model.requires_grad_(False)
        for n, m in model.named_modules():
            if n.endswith(".qkv"):
                m.requires_grad_(True)
        print(f"number of trainable parameters: {sum(p.numel() for p in peft_model.parameters() if p.requires_grad)}")

    peft_model.config.use_cache = False

    args = TrainingArguments(
        output_dir=model_save_path,
        per_device_train_batch_size=2,
        gradient_accumulation_steps=1,
        logging_steps=1,
        logging_first_step=True,
        num_train_epochs=5,
        save_steps=1000,
        learning_rate=2e-5,
        save_on_each_node=True,
        gradient_checkpointing=GRADIENT_CHECKPOINTING,
        report_to="none",
        bf16=True,
        max_grad_norm=1.0,
    )

    trainer = Trainer(
        model=peft_model,
        args=args,
        train_dataset=train_dataset,
        data_collator=DataCollatorForSeq2Seq(tokenizer=tokenizer, padding=True),
    )

    weights = {k: v.to(torch.float).clone() for k, v in peft_model.model.visual.state_dict().items() if 'qkv' in k}

    trainer.train()
    trainer.save_model(model_save_path + 'model')

    weights_ = {k: v.to(torch.float) for k, v in peft_model.model.visual.state_dict().items() if 'qkv' in k}
    for key in weights:
        print(f'{key}, {weights[key].equal(weights_[key])}')
```

With `USE_PEFT = False` and `GRADIENT_CHECKPOINTING = True`, I get 0 grad norms, but when I set `GRADIENT_CHECKPOINTING = False`, I do get grad norms > 0. Therefore, I think it's not caused by PEFT.

**Comment by githubnemo** [comment]
This is indeed a problem with transformers, thanks @BenjaminBossan for narrowing it down.

The problem is with `model.enable_input_require_grads()` - it doesn't seem to support visual language models yet and since the target module (`qkv`) is in the vision model but we only mark the language model input embeddings as `requires_grad=True`, gradient checkpointing fails.

There are two workarounds I can see:

- manually add the enable grad hook from `model.enable_input_require_grads`:
```python
    model.enable_input_require_grads()
    model.model.visual.patch_embed.register_forward_hook(lambda m, input, output: output.requires_grad_(True))
```
- use non-reentrant gradient checkpointing as it doesn't have the requirement for input gradients:
```
TrainingArguments(
    ...
    gradient_checkpointing_kwargs={'use_reentrant': False},
    ...
)
```

**Comment by Bytes-Lin** [comment]
> This is indeed a problem with transformers, thanks [@BenjaminBossan](https://github.com/BenjaminBossan) for narrowing it down.
> 
> The problem is with `model.enable_input_require_grads()` - it doesn't seem to support visual language models yet and since the target module (`qkv`) is in the vision model but we only mark the language model input embeddings as `requires_grad=True`, gradient checkpointing fails.
> 
> There are two workarounds I can see:
> 
>     * manually add the enable grad hook from `model.enable_input_require_grads`:
> 
> 
>     model.enable_input_require_grads()
>     model.model.visual.patch_embed.register_forward_hook(lambda m, input, output: output.requires_grad_(True))
> 
>     * use non-reentrant gradient checkpointing as it doesn't have the requirement for input gradients:
> 
> 
> ```
> TrainingArguments(
>     ...
>     gradient_checkpointing_kwargs={'use_reentrant': False},
>     ...
> )
> ```

Thanks for the attention and replying. 

I'm wondering whether transformers just take the input of LLM as input data and ignore the input data in ViT. If so, it's a potential bug for fine-tuning MLLMs.

**Comment by githubnemo** [comment]
> 
> I'm wondering whether transformers just take the input of LLM as input data and ignore the input data in ViT. If so, it's a potential bug for fine-tuning MLLMs.

I'm not sure if I understand you correctly but the issue is only with reentrant gradient checkpointing and has nothing to do with how inputs are passed down to the modules. I don't see an issue with MLLM except that `enable_input_require_grads` needs an extension to be more flexible with non-standard LLM.
